### PR TITLE
server : simplify state machine for slot

### PR DESCRIPTION
**Pull request by ngxson**
Currently, `state` of a slot is actually controlled by 2 variables: `slot.state` (2 possible values) and `slot.command` (3 possible values). This makes the total number of states become 2x3=6, but some of them are in fact invalid states.

This PR aims to simplify state machine of slot by unifying them into 4 states.

The state machine can be represent using the graph below:

```mermaid
graph TD;
    SLOT_STATE_IDLE-- new task -->SLOT_STATE_PROCESSING_PROMPT;
    SLOT_STATE_PROCESSING_PROMPT-- decode prompt -->SLOT_STATE_PROCESSING_PROMPT;
    SLOT_STATE_PROCESSING_PROMPT-- done processing prompt -->SLOT_STATE_DONE_PROMPT;
    SLOT_STATE_DONE_PROMPT-- is embedding -->SLOT_STATE_IDLE;
    SLOT_STATE_DONE_PROMPT-- is next-token prediction -->SLOT_STATE_GENERATING;
    SLOT_STATE_GENERATING-- decode next token -->SLOT_STATE_GENERATING;
    SLOT_STATE_GENERATING-- stop condition -->SLOT_STATE_IDLE;
```

---

- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [x] Low

### CODE REVIEW

**Comment by ngxson** [comment]
Hmm this seems to break context self-extend (passcode test - it's not run on CI but I run it locally)

**Comment by ngxson** [comment]
Benchmark with `-np 8 --http-threads 32` and `UVs = 16`

[master](https://github.com/ngxson/llama.cpp/tree/xsn/master_test_decode_count) (with `n_decode_total` metric added):

```
(waiting...)
```

PR:

```
# HELP llamacpp:n_decode_total Total number of llama_decode() calls
# TYPE llamacpp:n_decode_total counter
llamacpp:n_decode_total 3038
# HELP llamacpp:n_busy_slots_per_decode Average number of busy slots per llama_decode() call
# TYPE llamacpp:n_busy_slots_per_decode counter
llamacpp:n_busy_slots_per_decode 7.92956
# HELP llamacpp:prompt_tokens_seconds Average prompt throughput in tokens/s.
# TYPE llamacpp:prompt_tokens_seconds gauge
llamacpp:prompt_tokens_seconds 733.823
# HELP llamacpp:predicted_tokens_seconds Average generation throughput in tokens/s.
# TYPE llamacpp:predicted_tokens_seconds gauge
llamacpp:predicted_tokens_seconds 20.9693
```

**Comment by ngxson** [comment]
@ggerganov FYI, this PR also fixed the point about `notify_slot_changed()` that I brought up yesterday:

> I think `notify_slot_changed()` is probably a bottleneck. This function is used to bring back deferred tasks into main queue. However, because it is called inside `update_slots` where the slots are already set, this will make the slot to be idle during the current iteration, while in reality, the slot can take new task right away. (In other words, we bring back deferred tasks without processing them right away)

As you see on the benchmark above, `n_busy_slots_per_decode` does up from 7.7 to 7.9 with this PR. The consequence is that prompt processing tok/s increases (as prompts from new request appears more often in the batch), but the downside is that generation tok/s decreases. I'm testing on 1xA10G, but the result maybe different on better hardware.

**Comment by ngxson** [comment]
> Is it normal for the `input_tokens` to be different in the 2 tests?

No, it should not. I re-ran the test and I updated the result, so `input_tokens` is now matched between the 2 tests.

(I was having some network errors on the initial run, so that why it had less tokens)

**Review by slaren — COMMENTED** [review]
Looks good to me, but I don't really know the server code well enough to review this.

**Comment by ggerganov** [comment]
According to [040fdde](https://github.com/ggerganov/llama.cpp/pull/9283/commits/040fddee1cc8f219ba89c2369383b856f831031e) there was an address sanitizer issue. AFAICT the commit should not make a difference, except avoiding the copy of the task. Is the address sanitizer problem still present and how to reproduce?

**Comment by ngxson** [comment]
@ggerganov Thanks for pointing that out. You're right, indeed, there was a missing lock in `post(std::vector<server_task> & tasks)`. The function is added during the refactoring of multitask. This makes the CI to randomly fail, so I'll make a dedicated PR to fix that.

Beside, I changed `queue_tasks.erase(queue_tasks.begin())` to `queue_tasks.pop_front()` since queue is now std::deque. Not sure if it changes anything, but I re-ran the CI multiple times and it all passed:

- https://github.com/ngxson/llama.cpp/actions/runs/10735220528
- https://github.com/ngxson/llama.cpp/actions/runs/10735225928
- https://github.com/ngxson/llama.cpp/actions/runs/10735230159

File: examples/server/server.cpp [hunk 1]
```diff
@@ -451,14 +466,15 @@ struct server_queue {
...
-    void notify_slot_changed() {
-        // move deferred tasks back to main loop
+    // Call when the state of one slot is changed, it will move one task from deferred to main queue
+    void pop_deferred_task() {
         std::unique_lock<std::mutex> lock(mutex_tasks);
-        for (auto & task : queue_tasks_deferred) {
+        if (!queue_tasks_deferred.empty()) {
+            server_task task = queue_tasks_deferred.front();
+            queue_tasks_deferred.pop_front();
             queue_tasks.push_back(std::move(task));
```
Comment (ggerganov) [inline examples/server/server.cpp:475]: This should avoid a copy of `task`:

```suggestion
            queue_tasks.emplace_back(std::move(queue_tasks_deferred.front()));
            queue_tasks_deferred.pop_front();
```

**Review by ggerganov — APPROVED** [review]

**Comment by ngxson** [comment]
To make sure that it's not a fluke, I re-ran the CI 5 times:

image

So I can assume that this is safe to merge now.

**Pull request merged**